pkg/controller/provider/web/openstack/base.go

```diff
@@ -55,18 +56,33 @@ type PathBuilder struct {
 	// Database.
 	DB libmodel.DB
 	// Cached resources.
-	cache map[string]string
+	cache map[string]interface{}
 }
 
 // Build.
 func (r *PathBuilder) Path(m model.Model) (path string) {
 	var err error
 	if r.cache == nil {
-		r.cache = map[string]string{}
+		r.cache = map[string]interface{}{}
 	}
 	switch m.(type) {
+	case *model.Project:
+		project := m.(*model.Project)
+		path = project.Name
+		if project.IsDomain {
+			return
+		}
+		if project.ParentID != project.DomainID {
+			var parentProject *model.Project
+			parentProject, err = r.getProject(project.ParentID)
+			path = pathlib.Join(r.Path(parentProject), path)
+		}
+
 	case *model.VM:
-		path = m.(*model.VM).Name
+		vm := m.(*model.VM)
+		var project *model.Project
+		project, err = r.getProject(vm.TenantID)
+		path = pathlib.Join(r.Path(project), vm.Name)
```

Comment:
if we get an error above (err != nil) we'll probably fail here - is so, we'll need to handle the error here